backend/tests/test_event_creation.py

```diff
@@ -0,0 +1,372 @@
+    assert event.reason == "TestReason"
+    assert event.project is None
+    assert event.user_id == int(response.json()["id"])
+
+
+def test_user_deleted_cleanup(client, db, executor_name, unique_username):
+    executor = create_user(db, executor_name, Role.ADMIN)
+
+    response = client.post(
+        "/api/v1/users/",
+        json={"name": unique_username, "role": "user", "reason": reason},
+    )
+
+    assert response.status_code == 200
+    assert len(get_events_by_username(db, unique_username)) == 1
+    assert len(get_executed_events_by_user_id(db, executor.id)) == 1
+
+    user_id = int(response.json()["id"])
+    response = client.delete(f"/api/v1/users/{user_id}")
+
+    assert response.status_code == 204
+    assert len(get_events_by_username(db, unique_username)) == 0
+    assert len(get_executed_events_by_user_id(db, executor.id)) == 0
+
+
+@pytest.mark.parametrize(
+    "initial_role,target_role,expected_event_type",
+    [
+        (Role.USER, Role.ADMIN, EventType.ASSIGNED_ROLE_ADMIN),
+        (Role.ADMIN, Role.USER, EventType.ASSIGNED_ROLE_USER),
+    ],
+)
+def test_create_assign_user_role_event(
+    client,
+    db,
+    executor_name,
+    unique_username,
+    initial_role,
+    target_role,
+    expected_event_type,
+):
+    executor = create_user(db, executor_name, Role.ADMIN)
+    user = create_user(db, unique_username, initial_role)
+
+    response = client.patch(
+        f"/api/v1/users/{user.id}/roles",
+        json={"role": target_role.value, "reason": reason},
+    )
+
+    events = get_events_by_username(db, unique_username)
+
+    assert response.status_code == 200
+    assert len(events) == 1
+
+    event = events[0]
+
+    assert event.event_type == expected_event_type
+    assert event.executor_id == executor.id
+    assert event.reason == reason
+    assert event.project is None
+    assert event.user_id == user.id
+
+
+@pytest.mark.parametrize(
+    "permission,expected_permission_event_type",
+    [
+        (
+            ProjectUserPermission.READ,
+            EventType.ASSIGNED_PROJECT_PERMISSION_READ_ONLY,
+        ),
+        (
+            ProjectUserPermission.WRITE,
+            EventType.ASSIGNED_PROJECT_PERMISSION_READ_WRITE,
+        ),
+    ],
+)
+def test_create_user_added_to_project_event(
+    client,
+    db,
+    executor_name,
+    unique_username,
+    project,
+    permission,
+    expected_permission_event_type,
+):
+    executor = create_user(db, executor_name, Role.ADMIN)
+    user = create_user(db, unique_username, Role.USER)
+
+    response = client.post(
+        f"/api/v1/projects/{project.slug}/users/",
+        json={
+            "role": ProjectUserRole.USER.value,
+            "permission": permission.value,
+            "username": user.name,
+            "reason": reason,
+        },
+    )
+
+    assert response.status_code == 200
+
+    events = get_events_by_user_id(db, user.id)
+    assert len(events) == 3
+
+    for event, expected_event_type in zip(
```

Comment:
Try to avoid `for` loops and `if` statements in tests. You want tests to be fully deterministic and easy to read.

You can only test the most interesting fields, since the other fields have been covered in other tests as well.

Reply:
I also didn't like to use a loop here but I didn't find a cleaner solution to check all fields. But good point that I don't really have to check all the fields, so now I just check all fields only for the first event and for the other two I just check their event type.